### PR TITLE
fixes #1830 - auto assign puppet proxy if its not defined upon facts/reports event

### PR DESCRIPTION
**Pull request by ohadlevy**
an alternative to #940, also cover reports.

### CODE REVIEW

**Comment by ohadlevy** [comment]
all my tests pass locally, trying again [test]

**Comment by ohadlevy** [comment]
it helps if one tests in the right branch... fix incoming

**Comment by GregSutcliffe** [comment]
Yeah, this is what I was thinking of too, looks good. I'll try to give it a test in a while.

File: test/unit/host_test.rb [hunk 1]
```diff
@@ -892,6 +892,25 @@ class Host::Valid < Host::Base ; belongs_to :domain ; end
     assert_equal hosts.first.params['foo'], 'bar'
   end
 
+  test "should update puppet_proxy_id to the id of the validated proxy" do
+    sp  = SmartProxy.create!(:name => "Puppetmaster Proxy 2", :url => "http://another.one:4567") do |p|
```
Comment (domcleal) [inline test/unit/host_test.rb]: We already have a puppetmaster fixture for this?  (ditto below)

File: app/controllers/concerns/foreman/controller/smart_proxy_auth.rb [hunk 2]
```diff
@@ -56,13 +57,14 @@ def auth_smart_proxy(proxies = SmartProxy.all, require_cert = true)
...
-    logger.debug("Verifying request from #{request_hosts} against #{proxies.inspect}")
+    hosts = proxies.map { |p| URI.parse(p.url).host }.push(*Setting[:trusted_puppetmaster_hosts])
+    logger.debug("Verifying request from #{request_hosts} against #{hosts.inspect}")
 
-    unless proxies.detect { |p| request_hosts.include? p }
+    unless host = hosts.detect { |p| request_hosts.include? p }
       logger.warn "No smart proxy server found on #{request_hosts.inspect} and is not in trusted_puppetmaster_hosts"
       return false
     end
+    @detected_proxy = proxies[hosts.index(host)] if host
```
Comment (domcleal) [inline app/controllers/concerns/foreman/controller/smart_proxy_auth.rb]: Making hosts a hash to proxies would be slightly neater.

**Comment by domcleal** [comment]
:+1: generally, looks good.  We're missing functional tests though for the smart proxy auth behaviour, could you add some?

**Comment by domcleal** [comment]
Up to you too, but adding a "proxy=" argument to the import:hosts_and_facts rake task might be appreciated, just like the import:puppet_classes one.

**Comment by domcleal** [comment]
Merged as 5ab7985 for Foreman 1.4.0, thanks @ohadlevy and @witlessbird!

**Pull request closed without merging**